### PR TITLE
Add privacy messaging to AddExternalServicesPage

### LINKED ISSUE #11044: Improve privacy messaging before requesting addition of repository
Sourcegraph version:3.16
Platform information:MacOS, docker installed, latest chrome.'

#### Steps to reproduce:

Begin the process of adding a cost host at /site-admin/external-services/new.

Note there is no message to users about how Sourcegraph protects a user's repositories. 

#### Expected behavior:

Users are properly informed of the actions the software will take on their repositories.

#### Actual behavior:

New users to the platform do not yet trust Sourcegraph with their data, therefore, some users are reluctant to add a code host to their configuration. 

#### Potential solution
Inform the user what the software will do with the access provided. Suggested copy:

>How this install accesses your code host:
> • Clones the repositories you specify to create a local cache
> • Periodically pulls cloned repositories to ensure search results are current
> • Collects usage metrics such as number of users, user activity and number of searched repositories 
>     • We collect this data to ensure usage does not exceed free software requirements
> 
>This installation will not:
> • Send your code, queries, or other personal data to any online service. Your code and information are private.
> 

The message should be dismiss-able.

See the screenshot:

Privacy message

### PR DESCRIPTION
**Pull request by lguychard**
Fixes #11044

cc @rrhyne 

image

Also took the opportunity to convert `<SiteAdminAddExternalServicesPage />` to a function component.

### CODE REVIEW

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 1]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
+                        <li>
+                            Opening pull requests and syncing their metadata as part of{' '}
+                            <a
+                                href="https://docs.sourcegraph.com/user/campaigns"
+                                target="_blank"
+                                rel="noopener noreferrer"
+                            >
+                                code change campaigns
+                            </a>
+                            , if you have enabled this feature
```
Comment (lguychard) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:83]:
```suggestion
                            , if you have enabled this feature.
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 2]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
+                                href="https://docs.sourcegraph.com/user/campaigns"
+                                target="_blank"
+                                rel="noopener noreferrer"
+                            >
+                                code change campaigns
+                            </a>
+                            , if you have enabled this feature.
+                        </li>
+                    </ul>
+                    <div className="d-flex flex-row-reverse">
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:86]: Reversing the order seems a bit of a round-about way to accomplish alignment to the end of the row :)

```suggestion
                    <div className="d-flex align-items-end">
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 3]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
-    private getExternalServiceID(): string | null {
-        return new URLSearchParams(this.props.history.location.search).get('id') ?? null
+export const SiteAdminAddExternalServicesPage: React.FunctionComponent<Props> = props => {
+    const [hasDismissedPrivacyWarning, setHasDismissedPrivacyWarning] = useLocalStorage(
+        'HAS_DISMISSED_CODEHOST_PRIVACY_WARNING',
+        false
+    )
+    const dismissPrivacyWarning = useCallback(
+        (event: React.MouseEvent<HTMLButtonElement>) => {
+            event.preventDefault()
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:28]: What default is being prevented here? 🤔

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 4]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
+            setHasDismissedPrivacyWarning(true)
+        },
+        [setHasDismissedPrivacyWarning]
+    )
+    const id = new URLSearchParams(props.history.location.search).get('id') ?? null
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:33]: Noop expression
```suggestion
    const id = new URLSearchParams(props.history.location.search).get('id')
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 5]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
  */
-export class SiteAdminAddExternalServicesPage extends React.Component<Props> {
-    /**
-     * Returns the id of the external service from the URL parameters.
-     */
-    private getExternalServiceID(): string | null {
-        return new URLSearchParams(this.props.history.location.search).get('id') ?? null
+export const SiteAdminAddExternalServicesPage: React.FunctionComponent<Props> = props => {
+    const [hasDismissedPrivacyWarning, setHasDismissedPrivacyWarning] = useLocalStorage(
+        'HAS_DISMISSED_CODEHOST_PRIVACY_WARNING',
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:23]: Nit: usually use kebab-case or camelCase
```suggestion
        'hasDismissedCodeHostPrivacyWarning',
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 6]
```diff
@@ -17,45 +18,93 @@ interface Props extends ThemeProps {
...
+        false
+    )
+    const dismissPrivacyWarning = useCallback(
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx:26]: FYI I think `useCallback` is not _needed_ for DOM elements (our linting will detect that case and not complain I think)

**Pull request merged**